Penumbra/UI/SharedTagManager.cs

```diff
@@ -0,0 +1,160 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Dalamud.Interface;
+using ImGuiNET;
+using OtterGui;
+using OtterGui.Raii;
+using OtterGui.Widgets;
+using Penumbra.Mods;
+using Penumbra.UI.Classes;
+
+namespace Penumbra.UI;
+public sealed class SharedTagManager
+{
+    private static uint _tagButtonAddColor = ColorId.SharedTagAdd.Value();
+    private static uint _tagButtonRemoveColor = ColorId.SharedTagRemove.Value();
+
+    private static float _minTagButtonWidth = 15;
+
+    private const string PopupContext = "SharedTagsPopup";
+    private bool _isPopupOpen = false;
+
+
+    public IReadOnlyList<string> SharedTags { get; internal set; } = Array.Empty<string>();
+
+    public SharedTagManager()
+    {
+    }
+
+    public void ChangeSharedTag(int tagIdx, string tag)
+    {
+        if (tagIdx < 0 || tagIdx > SharedTags.Count)
+            return;
+
+        if (tagIdx == SharedTags.Count) // Adding a new tag
+        {
+            SharedTags = SharedTags.Append(tag).Distinct().Where(tag => tag.Length > 0).OrderBy(a => a).ToArray();
+        }
+        else // Editing an existing tag
+        {
+            var tmpTags = SharedTags.ToArray();
+            tmpTags[tagIdx] = tag;
+            SharedTags = tmpTags.Distinct().Where(tag => tag.Length > 0).OrderBy(a => a).ToArray();
+        }
+    }
+
+    public string DrawAddFromSharedTags(IReadOnlyCollection<string> localTags, IReadOnlyCollection<string> modTags, bool editLocal)
+    {
+        var tagToAdd = "";
+        if (ImGuiUtil.DrawDisabledButton(FontAwesomeIcon.Tags.ToIconString(), new Vector2(ImGui.GetFrameHeight()), "Add Shared Tag... (Right-click to close popup)",
+            false, true) || _isPopupOpen)
+            return DrawSharedTagsPopup(localTags, modTags, editLocal);
+
+
+        return tagToAdd;
+    }
+
+    private string DrawSharedTagsPopup(IReadOnlyCollection<string> localTags, IReadOnlyCollection<string> modTags, bool editLocal)
+    {
+        var selected = "";
+        if (!ImGui.IsPopupOpen(PopupContext))
```

Comment:
This doesn't look optimal, but I'll take a look after that first round of reviews when I compile and test it myself.
There's probably a nicer way to handle the popup.

Reply:
Yeah getting the popup to render and stick around was a bit of a struggle, and I'm not 100% happy with the behavior here (eg. clicking somewhere else on screen will move the popup there).

I would be open to any advice for how to better handle it.